File: src/dev/pci/SimplePciDevice.py
from m5.objects.PciDevice import PciDevice
from m5.params import *
from m5.SimObject import SimObject


class SimplePciDevice(PciDevice):
    type = "SimplePciDevice"
    cxx_header = "dev/pci/simple_pci_device.hh"
    cxx_class = "gem5::SimplePciDevice"

    pio_addr = Param.Addr(0x10000000, "PIO base address")
    pio_size = Param.Addr(0x1000, "PIO size")

File: src/dev/pci/simple_pci_device.cc
#include "dev/pci/simple_pci_device.hh"
#include "base/trace.hh"
#include "debug/SimplePciDevice.hh"

namespace gem5 {

SimplePciDevice::SimplePciDevice(const SimplePciDeviceParams &p)
    : PciDevice(p), pioAddr(p.pio_addr), pioSize(p.pio_size)
{
    // Allocate BAR0
    // config.space.config[PCI_BAR0] = pioAddr | PCI_BASE_ADDRESS_SPACE_IO;
}

Tick
SimplePciDevice::read(PacketPtr pkt)
{
    DPRINTF(SimplePciDevice, "Read request at offset %#x\n", pkt->getAddr());
    pkt->makeResponse();
    pkt->setLE<uint32_t>(0xDEADBEEF); // Dummy data
    return pioDelay;
}

Tick
SimplePciDevice::write(PacketPtr pkt)
{
    DPRINTF(SimplePciDevice, "Write request at offset %#x\n", pkt->getAddr());
    pkt->makeResponse();
    return pioDelay;
}

AddrRangeList
SimplePciDevice::getAddrRanges() const
{
    return { RangeSize(pioAddr, pioSize) };
}


// SimplePciDevice *
// SimplePciDeviceParams::create() const
// {
//     return new SimplePciDevice(*this);
// }

}   // namespace gem5
